src/valid/handles.rs

```diff
@@ -0,0 +1,631 @@
+        types: &UniqueArena<crate::Type>,
+    ) -> Result<(), InvalidHandleError> {
+        handle.check_valid_for_uniq(types).map(|_| ())
+    }
+
+    fn validate_constant_handle(
+        handle: Handle<crate::Constant>,
+        constants: &Arena<crate::Constant>,
+    ) -> Result<(), InvalidHandleError> {
+        handle.check_valid_for(constants).map(|_| ())
+    }
+
+    fn validate_expression_handle(
+        handle: Handle<crate::Expression>,
+        expressions: &Arena<crate::Expression>,
+    ) -> Result<(), InvalidHandleError> {
+        handle.check_valid_for(expressions).map(|_| ())
+    }
+
+    fn validate_function_handle(
+        handle: Handle<crate::Function>,
+        functions: &Arena<crate::Function>,
+    ) -> Result<(), InvalidHandleError> {
+        handle.check_valid_for(functions).map(|_| ())
+    }
+
+    fn validate_expression_handles(
+        (handle, expression): (Handle<crate::Expression>, &crate::Expression),
+        constants: &Arena<crate::Constant>,
+        types: &UniqueArena<crate::Type>,
+        local_variables: &Arena<crate::LocalVariable>,
+        global_variables: &Arena<crate::GlobalVariable>,
+        functions: &Arena<crate::Function>,
+    ) -> Result<(), InvalidHandleError> {
+        let validate_constant = |handle| Self::validate_constant_handle(handle, constants);
+        let validate_type = |handle| Self::validate_type_handle(handle, types);
+
+        match *expression {
+            crate::Expression::Access { base, index } => {
+                handle.check_dep(base)?.check_dep(index)?;
+            }
+            crate::Expression::AccessIndex { base, .. } => {
+                handle.check_dep(base)?;
+            }
+            crate::Expression::Constant(constant) => {
+                validate_constant(constant)?;
+            }
+            crate::Expression::Splat { value, .. } => {
+                handle.check_dep(value)?;
+            }
+            crate::Expression::Swizzle { vector, .. } => {
+                handle.check_dep(vector)?;
+            }
+            crate::Expression::Compose { ty, ref components } => {
+                validate_type(ty)?;
+                handle.check_dep_iter(components.iter().copied())?;
+            }
+            crate::Expression::FunctionArgument(_arg_idx) => (),
+            crate::Expression::GlobalVariable(global_variable) => {
+                global_variable.check_valid_for(global_variables)?;
+            }
+            crate::Expression::LocalVariable(local_variable) => {
+                local_variable.check_valid_for(local_variables)?;
+            }
+            crate::Expression::Load { pointer } => {
+                handle.check_dep(pointer)?;
+            }
+            crate::Expression::ImageSample {
+                image,
+                sampler,
+                gather: _,
+                coordinate,
+                array_index,
+                offset,
+                level: _,
```

Comment:
The `level` field is a `SampleLevel`, which can have handles.

Reply:
Gonna need to make some pizza…with all this d'oh. 😩

Resolved with e1229f1b.